### PR TITLE
add LDL loss

### PR DESCRIPTION
**Pull request by csjliang**

### CODE REVIEW

File: basicsr/losses/loss_util.py [hunk 1]
```diff
@@ -93,3 +94,31 @@ def wrapper(pred, target, weight=None, reduction='mean', **kwargs):
         return loss
 
     return wrapper
+
+
+def get_local_weights(residual, ksize):
+
```
Comment (xinntao) [inline basicsr/losses/loss_util.py:100]: Since this function is only used by `get_refined_artifact_map`, it is better to add a docstring to specify it.

File: basicsr/models/realesrgan_model.py [hunk 2]
```diff
@@ -207,6 +208,8 @@ def optimize_parameters(self, current_iter):
 
         self.optimizer_g.zero_grad()
         self.output = self.net_g(self.lq)
+        if self.cri_artifacts:
```
Comment (xinntao) [inline basicsr/models/realesrgan_model.py:211]: I suggest to use self.cri_ldl, as ldl is the method name of your paper. So using ldl will let readers understand it quickly.

File: basicsr/losses/loss_util.py [hunk 3]
```diff
@@ -93,3 +94,31 @@ def wrapper(pred, target, weight=None, reduction='mean', **kwargs):
...
+    return pixel_level_weight
+
+
+def get_refined_artifact_map(img_gt, img_output, img_ema, ksize):
+    """Calculate the artifact map of LDL
+    (Details or Artifacts: A Locally Discriminative Learning Approach to Realistic Image Super-Resolution. In CVPR 2022)
+    """
+
+    residual_ema = torch.sum(torch.abs(img_gt - img_ema), 1, keepdim=True)
+    residual_SR = torch.sum(torch.abs(img_gt - img_output), 1, keepdim=True)
```
Comment (xinntao) [inline basicsr/losses/loss_util.py:116]: resudal_sr

it is a convention, for variable name, we use lowercase.

File: basicsr/models/realesrgan_model.py [hunk 4]
```diff
@@ -216,6 +219,12 @@ def optimize_parameters(self, current_iter):
                 l_g_pix = self.cri_pix(self.output, l1_gt)
                 l_g_total += l_g_pix
                 loss_dict['l_g_pix'] = l_g_pix
+            if self.cri_artifacts:
+                pixel_weight = get_refined_artifact_map(self.gt, self.output, self.output_ema, 7)
+                l_g_artifacts = self.cri_artifacts(
```
Comment (xinntao) [inline basicsr/models/realesrgan_model.py:224]: suggest to use `l_g_ldl`

File: basicsr/models/srgan_model.py [hunk 5]
```diff
@@ -51,6 +51,11 @@ def init_training_settings(self):
         else:
             self.cri_pix = None
 
+        if train_opt.get('artifacts_opt'):
```
Comment (xinntao) [inline basicsr/models/srgan_model.py:54]: Add a comment here, what is 'artifacts_opt' 

also suggest to use `ldl_opt`

File: options/train/LDL/train_LDL_Real_x4.yml [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+path:
+  # use the pre-trained Real-ESRNet model
+  pretrain_network_g: experiments/pretrained_models/RealESRGAN/RealESRNet_x4plus.pth
+  param_key_g: params_ema
+  strict_load_g: true
+  resume_state: ~
+
+# training settings
+train:
+  optim_g:
```
Comment (xinntao) [inline options/train/LDL/train_LDL_Real_x4.yml:108]: put  `ema_decay` here, so keep consistent with RealESRGAN settings,

File: options/train/LDL/train_LDL_Real_x4.yml [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+  wandb:
+    project: ~
+    resume_id: ~
+
+# dist training settings
+dist_params:
+  backend: nccl
+  port: 29500
+
+# run
```
Comment (xinntao) [inline options/train/LDL/train_LDL_Real_x4.yml:184]: Remove it. 

OR 
I think it may be better to add such running commands on the top of each file.

**Comment by csjliang** [comment]
Hi, thanks for your suggestion. I have updated the codes accordingly.
